build: debug enable addrman consistency checks

willcl-ark opened the pull request:
The `--enable-debug` flag should default to enabling debug consistency checks. This PR enables addrman and mempool consistency checks at a rate of once per 1000 operations, by default when running in debug mode.

To opt out of the checks when running a debug build specify the runtime `-checkmempool=0` and `-checkaddrman=0` debug options, as described in developer-notes.md

On my machine, configuring these checks to run every 1000 operations saw them run no more than once per minute (and often much less frequently), which did not introduce any significant overhead.

This PR does not include a `-checkblockindex` default consistency check for mainnet when running in debug mode.

fixes #24709

willcl-ark committed:
doc: Add DDEBUG_ADDRMAN to developer-notes.md

maflcko reviewed (commented):
Concept ACK.

Though, I haven't checked how expensive those checks are. If they make the program unusable slow, I can see that someone would Nack this being on by default.

What about mempool checks?

mzumsande reviewed (commented):
Concept ACK - I've run with `-checkaddrman=1` before, never found bitcoind to be unusably slow. Not sure about the mempool checks though.

Generally speaking, I think it's a bit  strange that `--enable-debug` mixes two different concepts:
1.) Compiler flags for attaching a debugger
2.) More (but costly) runtime checks (like `-DDEBUG_LOCKORDER` currently, plus additionally custom checks for addrman / mempool etc.)

I don't think it's very common that people would want both 1 and 2 at the same time:
If I want to attach a debugger for better analysis of a specific part of the code, I'm not interested in checks for unrelated parts -  if I want to run with more checks to find possible unknown bugs, I'm not interested in attaching a debugger.
So why do we have one flag for the two?

willcl-ark commented:
@mzumsande FYI this PR is equivalent to running with `-checkaddrman=1` which does use quite some resources as it does a full addrman consistency check on pretty much every addrman operation (Add, Good, Attempt, Get, Select, Connected etc.).

For me this manifests as saturating an entire core via a combination of the `b-msghand`, `b-net` and `b-opencon` threads as these all spawn consistency checks. My `bitcoind` process  usually idles at ~5%, so to have one core @ 100% is a pretty big change, but my machine has 8 physical cores so it's also not too terrible...

I have not profiled with perf as it is so obvious in top.

I also agree that there are two debug scenarios (which are already mixed) being further mixed by this PR, i) enabling debugging symbols and ii) running the application in "debug mode". 

In a vacuum I would agree with @MarcoFalke in introducing an `--enable-debug-symbols` flag, however in this project, with `--enable-debug` as the encumbent and there being a lot of developer expectation around what it does, perhaps the new flag  should be `--debug-mode` which both implies `--enable-debug` and actually runs with all these debug mode options enabled by default?

ajtowns commented on src/addrman.h:
```diff
@@ -27,7 +27,11 @@ class InvalidAddrManVersionError : public std::ios_base::failure
 
 /** Default for -checkaddrman */
+#ifdef DEBUG_ADDRMAN
+static constexpr int32_t DEFAULT_ADDRMAN_CONSISTENCY_CHECKS{1};
```
If these checks are going to be enabled by default on some builds, shouldn't the default be much higher; eg 100 or 1000? Doing it on every call is useful for live debugging or small scale tests, but seems overkill for a node's runtime default?

willcl-ark replied:
Yes I agree, the performance hit is too much for any kind of normal operation.

I updated the branch with @MarcoFalke's suggestion along with a new debug option (which I called `--enable-debug-checks`, and not `--debug-slow`).

I'll run a few runtime tests on various values for the checks to see if I can find a sweet spot for the frequency, then update the doc commit and push an update here.

willcl-ark committed:
build: enable addrman consistency checks in debug

When configured with --enable-debug run addrman consistency checks every
1000 operations by default.

Running every 1000 operations minimises the CPU overhead on the node and
keeps it viable to run a node in debug mode on mainnet.

willcl-ark committed:
build: enable mempool consistency checks in debug

When configured with --enable-debug run mempool consistency checks every
1000 operations by default.

willcl-ark committed:
doc: Add DDEBUG_MEMPOOL to developer-notes.md

willcl-ark commented:
Nope, I also agree with the conclusions from AJ in 24709 about just setting this as a runtime option, so will close this.
